### PR TITLE
chore(logstreams): zero-copy LogStreamReaderImpl

### LINKED ISSUE #3473: Refactor slow BufferedLogStreamReader implementation
**Description**

Since we've moved to using the Atomix log, we currently read entries as blocks. This means for every entry, Kryo will allocate a fresh buffer for that entry, and we copy that buffer into the BufferedLogStreamReader's read buffer. Since each entry is a block, we don't perform read-ahead anymore and gain no benefit from batching. Additionally, the copy here is completely unnecessary, and when dealing with large events, can decrease performance by as much as 50% in micro-benchmarks (YMMV).

There are two options here:

1. Keep the reader as a read-ahead, buffered reader, in which case we should not read a single entry but read X of them at once - there we get some advantage of doing read-ahead on the normal scan behaviour (common case), and the copy here makes sense.
1. Keep the single entry reader for now and remove the current unnecessary copy/buffer allocation, use the Kryo provided buffer directly (since it is not reused anywhere else). In my own micro-benchmarks, this greatly improved performance.

I would lean towards the latter since it increases performance quickly, is cheap, and we can always add read-ahead capabilities later on (but I would revisit this when we invest time in alternatives to Kryo/re-investigate using memory-mapped files).

@Zelldon let me know what you think :eyes: 

EDIT: you can use the following tests to sort of bench mark things; it's not perfect by any means, or even as good as JMH, but by running it forever you can get some idea of the performance difference with/without copy:

In `LogStreamReaderTest.java`:
```java
@Test
  public void shouldScanLogFast() {
    // given
    final var eventCount = 1000;
    final var watch = Stopwatch.createUnstarted();
    IntStream.range(0, 30).forEach(i -> {
      writer.writeEvents(eventCount, BIG_EVENT_VALUE);
      LockSupport.parkNanos(1000 * 200);
    });

    // when
    reader.seekToFirstEvent();
    watch.start();
    while (reader.hasNext()) {
      reader.next();
    }

    // then
    Loggers.LOGSTREAMS_LOGGER.debug("Elapsed: {}", watch.stop());
  }
```

Another point hinting that the copy makes things slower is that everything slows down proportionally to the size of the events, not so much the amount of events.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

- simplifies the `LogStreamReader` implementation, removing the "buffered" part and any unnecessary copy; this is safe as for now, Kryo will allocate a new buffer for every deserialized entry
- fixes edge case when reading the last log entry if it is not a ZeebeEntry

## Related issues

Closes #3473 

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by npepinpe** [comment]
I'll let you merge your PR and rebase after ;)

Also this is related to #3473 but I'd like to keep it open for now so I didn't add it to the PR (since bors would later close it).

**Comment by npepinpe** [comment]
Also before we merge I'd like to run a benchmark :+1:

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: Oh this wasn't closed before :eyes:

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: Do we need to do this for the other buffer as well?

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: How we can say this?

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: I think this NOT_COMMITED state is never set right?

File: logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamReader.java
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamReader.java]: Why we should ever want this :laughing:

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Cool :+1: just some smaller things

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: Hm, on one hand if we don't it means we keep a reference to the allocated buffer (but I hope the reader is also garbage collected if it is closed, so not too bad). If we do, however, then if someone is still using the `returnedEvent` then any usage will explode. So as it is now it means you can still use the `returnedEvent` after you close the reader - the `returnedEvent` is only made unusable when you do a `next()` call.

I think this is the behaviour I would expect? But maybe it's safer to reset it...not sure. What do you think?

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: That was already like this :sweat_smile: I think the logic was that since we can __only__ read committed events, we assume they are always committed? Actually I think we can remove this since it's useless now

Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java]: Yes lets remove it

File: logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamReader.java
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamReader.java]: Good question :sweat_smile: I think this is called for deletion? :man_shrugging:

Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/log/LogStreamReader.java]: Should probably not be part of the public api :laughing: but we refactor this later

**Review by ChrisKujawa — APPROVED** [review]

**Pull request merged**
